"contrib/run_function.sh: line 16: kill: (3891525) - No such process" when make strat-kind

<!--- Provide a general summary of the issue in the Title above -->
I am trying to develop the provider, but some errors occur when creating the kind cluster.

## Expected Behaviour
<!--- If you're describing a bug, tell us what should happen -->
Kind cluster should run OK.
<!--- If you're suggesting a change/improvement, tell us how it should work -->

## Current Behaviour
```shell
>>> Creating Kubernetes v1.18.8 cluster kind
Creating cluster "kind" ...
DEBUG: docker/images.go:58] Image: kindest/node:v1.18.8 present locally
 ✓ Ensuring node image (kindest/node:v1.18.8) 🖼 
 ✓ Preparing nodes 📦  
 ✓ Writing configuration 📜 
 ✓ Starting control-plane 🕹️ 
 ✓ Installing CNI 🔌 
 ✓ Installing StorageClass 💾 
 ✓ Waiting ≤ 5m0s for control-plane = Ready ⏳ 
 • Ready after 25s 💚
Set kubectl context to "kind-kind"
You can now use your cluster with:

kubectl cluster-info --context kind-kind

Thanks for using kind! 😊
>>> Waiting for CoreDNS
Waiting for deployment "coredns" rollout to finish: 0 of 2 updated replicas are available...
Waiting for deployment "coredns" rollout to finish: 1 of 2 updated replicas are available...
deployment "coredns" successfully rolled out
Applying namespaces
namespace/openfaas created
namespace/openfaas-fn created
secret/basic-auth created
Waiting for helm install to complete.
Release "openfaas" does not exist. Installing it now.
NAME: openfaas
LAST DEPLOYED: Thu May 20 14:42:39 2021
NAMESPACE: openfaas
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
To verify that openfaas has started, run:

  kubectl -n openfaas get deployments -l "release=openfaas, app=openfaas"
Waiting for deployment "prometheus" rollout to finish: 0 out of 1 new replicas have been updated...
Waiting for deployment "prometheus" rollout to finish: 0 of 1 updated replicas are available...
deployment "prometheus" successfully rolled out
Waiting for deployment "gateway" rollout to finish: 0 of 1 updated replicas are available...
deployment "gateway" successfully rolled out
deployment "gateway" successfully rolled out
contrib/run_function.sh: line 16: kill: (3891525) - No such process
make: *** [Makefile:64: start-kind] Error 1
```
<!--- If describing a bug, tell us what happens instead of the expected behavior -->
<!--- If suggesting a change/improvement, explain the difference from current behavior -->

## Are you a GitHub Sponsor (Yes/No?)
<!--- Given this request for help, how are you supporting the project? -->
No

Check at: https://github.com/sponsors/openfaas
- [x] Yes
- [ ] No

## List All Possible Solutions and Workarounds
<!--- Suggest a fix/reason for the bug, or ideas how to implement  -->
<!--- the addition or change -->
<!--- Is there a workaround which could avoid making changes? -->

## Which Solution Do You Recommend?
<!--- Pick your preferred solution, if you were to implement and maintain this change -->
Not yet
## Steps to Reproduce (for bugs)
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->
1.
2.
3.
4.

## Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* FaaS-CLI version ( Full output from: `faas-cli version` ):
```shell
CLI:
 commit:  2cec97955a254358de5443987bedf8ceee272cf8
 version: 0.13.9

Gateway
 uri:     http://192.168.1.19:31112
 version: 0.20.12
 sha:     a6dbb4cd0285f6dbc0bc3f43f72ceacdbdf6f227


Provider
 name:          faas-netes
 orchestration: kubernetes
 version:       0.13.2 
 sha:           6caf24ba21f7af171e2ade40daa033b05584b695
```
* Docker version `docker version` (e.g. Docker 17.0.05 ):
```shell
Client: Docker Engine - Community
 Version:           20.10.6
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        370c289
 Built:             Fri Apr  9 22:44:36 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.6
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       8728dd2
  Built:            Fri Apr  9 22:43:02 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.4
  GitCommit:        05f951a3781f4f2c1911b05e61c160e9c30eaa8e
 runc:
  Version:          1.0.0-rc93
  GitCommit:        12644e614e25b05da6fd08a38ffa0cfe1903fdec
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```
* What version and distriubtion of Kubernetes are you using? `kubectl version`
```shell
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-12T14:18:45Z", GoVersion:"go1.16.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.8", GitCommit:"9f2892aab98fe339f3bd70e3c470144299398ace", GitTreeState:"clean", BuildDate:"2020-09-14T07:44:34Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
WARNING: version difference between client (1.21) and server (1.18) exceeds the supported minor version skew of +/-1
```
* Operating System and version (e.g. Linux, Windows, MacOS):
Linux/Centos 8
* Link to your project or a code example to reproduce issue:

* What network driver are you using and what CIDR? i.e. Weave net / Flannel
calico

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -50,8 +50,6 @@ You can use `OF_DEV_ENV` to set a custom name for the cluster. The default value
 As you are developing on `faas-netes`, you will want to build and test your own local images.  This can easily be done using the following commands
 
 ```sh
-export KUBECONFIG="$(kind get kubeconfig-path --name="${OF_DEV_ENV:-kind}")"
-
 make build
 
 kind load docker-image --name="${OF_DEV_ENV:-kind}" openfaas/faas-netes:latest

--- a/Makefile
+++ b/Makefile
@@ -61,12 +61,7 @@ render-charts:
 	./contrib/create-static-manifest.sh ./chart/openfaas ./yaml_armhf ./chart/openfaas/values-armhf.yaml
 
 start-kind: ## attempt to start a new dev environment
-	@./contrib/create_dev.sh \
-		&& echo "" \
-		&& printf '%-15s:\t %s\n' 'Web UI' 'http://localhost:31112/ui' \
-		&& printf '%-15s:\t %s\n' 'User' 'admin' \
-		&& printf '%-15s:\t %s\n' 'Password' $(shell cat password.txt) \
-		&& printf '%-15s:\t %s\n' 'CLI Login' "faas-cli login --username admin --password $(shell cat password.txt)"
+	@./contrib/create_dev.sh
 
 stop-kind: ## attempt to stop the dev environment
 	@./contrib/stop_dev.sh

--- a/contrib/create_cluster.sh
+++ b/contrib/create_cluster.sh
@@ -3,9 +3,25 @@
 DEVENV=${OF_DEV_ENV:-kind}
 KUBE_VERSION=v1.18.8
 
-echo ">>> Creating Kubernetes ${KUBE_VERSION} cluster ${DEVENV}"
+EXISTS=$(kind get clusters | grep "$DEVENV")
 
-kind create cluster --wait 5m --image kindest/node:${KUBE_VERSION} --name "$DEVENV" -v 1
+if [ "$EXISTS" = "$DEVENV" ]; then
+    while true; do
+    read -p "Kind cluster '$DEVENV' already exists, do you want to continue with this cluster? (yes/no) " yn
+    case $yn in
+        [Yy]* ) break;;
+        [Nn]* ) exit 1;;
+        * ) echo "Please answer yes or no.";;
+    esac
+done
+fi
+
+
+if [ "$EXISTS" != "$DEVENV" ]; then
+    echo ">>> Creating Kubernetes ${KUBE_VERSION} cluster ${DEVENV}"
+
+    kind create cluster --wait 5m --image kindest/node:${KUBE_VERSION} --name "$DEVENV" -v 1
+fi
 
 echo ">>> Waiting for CoreDNS"
 kubectl --context "kind-$DEVENV" -n kube-system rollout status deployment/coredns

--- a/contrib/create_dev.sh
+++ b/contrib/create_dev.sh
@@ -2,6 +2,16 @@
 
 # Creates an installation of OpenFaaS in kind for development
 
-contrib/create_cluster.sh
+set -e
+contrib/create_cluster.sh || exit 0;
 contrib/deploy.sh
 contrib/run_function.sh
+
+cd "$(git rev-parse --show-toplevel)"
+echo ""
+echo ""
+echo "Local dev cluster details:"
+printf '%-10s:\t %s\n' 'Web UI' 'http://localhost:8080/ui'
+printf '%-10s:\t %s\n' 'User' 'admin'
+printf '%-10s:\t %s\n' 'Password' "$(cat password.txt)"
+printf '%-10s:\t %s\n' 'CLI Login' "faas-cli login --username admin --password $(cat password.txt)"

--- a/contrib/deploy.sh
+++ b/contrib/deploy.sh
@@ -5,6 +5,7 @@ set -e
 DEVENV=${OF_DEV_ENV:-kind}
 OPERATOR=${OPERATOR:-0}
 
+echo ""
 echo "Applying namespaces"
 kubectl --context "kind-$DEVENV" apply -f ./namespaces.yml
 
@@ -17,18 +18,25 @@ if [ -x "$(command -v $sha_cmd)" ]; then
     sha_cmd="shasum"
 fi
 
-PASSWORD=$(head -c 16 /dev/urandom| $sha_cmd | cut -d " " -f 1)
-echo -n $PASSWORD > password.txt
+PASSWORD_EXISTS=$(kubectl get secret -n openfaas basic-auth | wc -c)
+if [ $PASSWORD_EXISTS -eq 0 ]; then
+    PASSWORD=$(head -c 16 /dev/urandom| $sha_cmd | cut -d " " -f 1)
+    echo -n $PASSWORD > password.txt
 
-kubectl --context "kind-$DEVENV" -n openfaas create secret generic basic-auth \
---from-literal=basic-auth-user=admin \
---from-literal=basic-auth-password="$PASSWORD"
+    kubectl --context "kind-$DEVENV" -n openfaas create secret generic basic-auth \
+    --from-literal=basic-auth-user=admin \
+    --from-literal=basic-auth-password="$PASSWORD"
+else
+    PASSWORD=$(kubectl get secret -n openfaas basic-auth -o=go-template='{{index .data "basic-auth-password"}}' | base64 --decode)
+    echo -n $PASSWORD > password.txt
+fi
 
 CREATE_OPERATOR=false
 if [ "${OPERATOR}" == "1" ]; then
     CREATE_OPERATOR="true"
 fi
 
+echo ""
 echo "Waiting for helm install to complete."
 
 helm upgrade \

--- a/contrib/run_function.sh
+++ b/contrib/run_function.sh
@@ -13,21 +13,30 @@ then
 fi
 
 if [ -f "of_${DEVENV}_portforward.pid" ]; then
-    kill $(<of_${DEVENV}_portforward.pid)
+    kill "$(<of_${DEVENV}_portforward.pid)" > /dev/null 2> /dev/null || :
 fi
 
 # quietly start portforward and put it in the background, it will not
 # print every connection handled
-kubectl --context "kind-$DEVENV" port-forward deploy/gateway -n openfaas 31112:8080 &>/dev/null & \
+kubectl --context "kind-$DEVENV" port-forward deploy/gateway -n openfaas 8080:8080 &>/dev/null & \
     echo -n "$!" > "of_${DEVENV}_portforward.pid"
 
 # port-forward needs some time to start
 sleep 10
 
-export OPENFAAS_URL=http://127.0.0.1:31112
+export OPENFAAS_URL=http://127.0.0.1:8080
 
 # Login into the gateway
+echo ""
+echo "Login to the gateway..."
 cat ./password.txt | faas-cli login --username admin --password-stdin
+
+IS_DEPLOYED=$(faas-cli list | grep echo | wc -c)
+if [ $IS_DEPLOYED -gt 0 ];then
+    echo ""
+    echo "test function 'echo' already deployed"
+    exit;
+fi
 
 # Deploy via the REST API which can test either the controller or operator
 faas-cli deploy --image=functions/alpine:latest --fprocess=cat --name "echo"

